### ISSUE TITLE
[no-floating-promises] ignoreVoid=true doesn't work in Array's forEach, in setTimeout, in optional chaining

### ISSUE THREAD

**Issue by dko-slapdash** [typescript-eslint/typescript-eslint#4018]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", { ignoreVoid: false }],
  }
}
```

```TS
const func: () => Promise<number> = async () => 10;
const funcOptional: (() => Promise<number>) | undefined = async () => 10;

[1, 2].forEach(() => void func()); // <-- does NOT raise an error, although it should
setTimeout(() => void func(), 10); // <-- does NOT raise an error, although it should
funcOptional?.(); // <-- does NOT raise an error, although it should

void func(); // <-- raises an error which is what we want
func(); // <-- raises an error which is what we want
funcOptional!(); // <-- raises an error which is what we want
```

**Expected Result**

The expectation is that `void` is disallowed for Promises in ALL contexts, not only in some of them.

**Actual Result**

It doesn't complain.

**Versions**

| package                                   | version |
| ----------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.0.0` |
| `@typescript-eslint/parser`               | `5.0.0` |
| `TypeScript`                              | `4.4.4` |
| `ESLint`                                  | `8.0.1` |
| `node`                                    | `16.11.1` |

**Comment by bradzacher** [comment]
You're misunderstanding what `no-floating-promises` does.

It only checks promises at the top level, as a statement - where you can "create" a promise and leave it unhandled by anything - causing a "floating" promise.

If you pass a promise to a function or return it from a function then it's not floating as the assumption is that the caller / callee will handle it.
